File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportC#/Persistence/ado/SeatDbRepository.cs
using FirmaTransportC_.Model;
using FirmaTransportC_.Persistence.utils;
using FirmaTransportC_.Persistence;
using log4net;
using System;
using System.Collections.Generic;
using System.Data;

namespace FirmaTransportC_.Persistence.ado
{
    public class SeatDbRepository : ISeatRepository
    {
        private static readonly ILog log = LogManager.GetLogger("SeatDbRepository");
        private readonly IDictionary<string, string> props;

        public SeatDbRepository(IDictionary<string, string> props)
        {
            log.Info("Creating SeatDbRepository with properties: " + string.Join(", ", props));
            this.props = props;
        }

        public long Add(Seat entity)
        {
            log.InfoFormat("Adding seat: {0}", entity);
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "INSERT INTO seats (number, reservation_id, trip_id, clientName) VALUES (@number, @reservationId, @tripId, @clientName); SELECT last_insert_rowid();";
                command.Parameters.Add(CreateParam(command, "@number", entity.Number));
                command.Parameters.Add(CreateParam(command, "@reservationId", entity.ReservationId.HasValue ? entity.ReservationId.Value : DBNull.Value));
                command.Parameters.Add(CreateParam(command, "@tripId", entity.TripId));
                command.Parameters.Add(CreateParam(command, "@clientName", entity.ClientName ?? "-"));
                var result = command.ExecuteScalar();
                if (result != null)
                {
                    long id = Convert.ToInt64(result);
                    entity.Id = id;
                    return id;
                }
            }
            catch (Exception ex)
            {
                log.Error("Error adding seat", ex);
            }
            return -1;
        }

        public void Delete(long id)
        {
            log.InfoFormat("Deleting seat with id: {0}", id);
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "DELETE FROM seats WHERE id = @id";
                command.Parameters.Add(CreateParam(command, "@id", id));
                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error deleting seat", ex);
            }
        }

        public void Update(Seat entity)
        {
            log.InfoFormat("Updating seat: {0}", entity);
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "UPDATE seats SET number=@number, reservation_id=@reservationId, trip_id=@tripId, client_name=@clientName WHERE id=@id";
                command.Parameters.Add(CreateParam(command, "@number", entity.Number));
                command.Parameters.Add(CreateParam(command, "@reservationId", entity.ReservationId.HasValue ? entity.ReservationId.Value : DBNull.Value));
                command.Parameters.Add(CreateParam(command, "@tripId", entity.TripId));
                command.Parameters.Add(CreateParam(command, "@clientName", entity.ClientName ?? "-"));
                command.Parameters.Add(CreateParam(command, "@id", entity.Id));
                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error updating seat", ex);
            }
        }

        public Seat FindOne(long id)
        {
            log.InfoFormat("Finding seat with id: {0}", id);
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "SELECT * FROM seats WHERE id=@id";
                command.Parameters.Add(CreateParam(command, "@id", id));
                using var reader = command.ExecuteReader();
                if (reader.Read())
                    return ExtractSeat(reader);
            }
            catch (Exception ex)
            {
                log.Error("Error finding seat", ex);
            }
            return null;
        }

        public List<Seat> FindAll()
        {
            log.Info("Finding all seats");
            var seats = new List<Seat>();
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "SELECT * FROM seats";
                using var reader = command.ExecuteReader();
                while (reader.Read())
                    seats.Add(ExtractSeat(reader));
            }
            catch (Exception ex)
            {
                log.Error("Error finding seats", ex);
            }
            return seats;
        }

        public List<Seat> FindSeatsForTrip(long tripId)
        {
            log.InfoFormat("Finding seats for trip: {0}", tripId);
            var seats = new List<Seat>();
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "SELECT * FROM seats WHERE trip_id=@tripId";
                command.Parameters.Add(CreateParam(command, "@tripId", tripId));
                using var reader = command.ExecuteReader();
                while (reader.Read())
                    seats.Add(ExtractSeat(reader));
            }
            catch (Exception ex)
            {
                log.Error("Error finding seats for trip", ex);
            }
            return seats;
        }

        public List<Seat> FindSeatsForReservation(long reservationId)
        {
            log.InfoFormat("Finding seats for reservation: {0}", reservationId);
            var seats = new List<Seat>();
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "SELECT * FROM seats WHERE reservation_id=@reservationId";
                command.Parameters.Add(CreateParam(command, "@reservationId", reservationId));
                using var reader = command.ExecuteReader();
                while (reader.Read())
                    seats.Add(ExtractSeat(reader));
            }
            catch (Exception ex)
            {
                log.Error("Error finding seats for reservation", ex);
            }
            return seats;
        }

        public int CountAvailableSeatsForTrip(long tripId)
        {
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "SELECT COUNT(*) FROM seats WHERE trip_id=@tripId AND reservation_id IS NULL";
                command.Parameters.Add(CreateParam(command, "@tripId", tripId));
                return Convert.ToInt32(command.ExecuteScalar());
            }
            catch (Exception ex)
            {
                log.Error("Error counting available seats", ex);
            }
            return -1;
        }

        private Seat ExtractSeat(IDataReader reader)
        {
            long id = reader.GetInt64(0);
            int number = reader.GetInt32(1);
            long? reservationId = reader.IsDBNull(2) ? null : reader.GetInt64(2);
            long tripId = reader.GetInt64(3);
            string clientName = reader.IsDBNull(4) ? "-" : reader.GetString(4);
            return new Seat(id, number, reservationId, tripId, clientName);
        }

        private IDbDataParameter CreateParam(IDbCommand cmd, string name, object value)
        {
            var param = cmd.CreateParameter();
            param.ParameterName = name;
            param.Value = value;
            return param;
        }
    }
}

File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportJava/Persistence/src/main/java/transport/persistence/repository/jdbc/SeatDbRepository.java
package transport.persistence.repository.jdbc;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import transport.model.Seat;
import transport.persistence.repository.SeatRepository;
import transport.persistence.repository.utils.JdbcUtils;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class SeatDbRepository implements SeatRepository {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger(SeatDbRepository.class);

    public SeatDbRepository(Properties properties) {
        logger.info("Initializing SeatDbRepository with properties: {}", properties);
        dbUtils = new JdbcUtils(properties);
    }

    @Override
    public List<Seat> findSeatsForTrip(Long tripId) {
        logger.info("Finding seats for trip with id: {}", tripId);

        String sql = "SELECT * FROM seats WHERE trip_id = ?";
        List<Seat> seats = new ArrayList<>();

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setLong(1, tripId);

            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    seats.add(buildSeatFromResultSet(result));
                }
            }

        } catch (Exception e) {
            logger.error("Error finding seats for trip {}", tripId, e);
        }

        return seats;
    }

    @Override
    public List<Seat> findSeatsForReservation(Long reservationId) {
        logger.info("Finding seats for reservation with id: {}", reservationId);

        String sql = "SELECT * FROM seats WHERE reservation_id = ?";
        List<Seat> seats = new ArrayList<>();

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setLong(1, reservationId);

            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    seats.add(buildSeatFromResultSet(result));
                }
            }

        } catch (Exception e) {
            logger.error("Error finding seats for reservation {}", reservationId, e);
        }

        return seats;
    }

    @Override
    public int countAvailableSeatsForTrip(Long tripId) {
        logger.info("Finding available seats for trip with id: {}", tripId);
        String sql = "SELECT COUNT(*) AS available_seats FROM seats WHERE trip_id = ? AND reservation_id IS NULL";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setLong(1, tripId);

            try (ResultSet result = statement.executeQuery()) {
                if (result.next()) {
                    int availableSeats = result.getInt("available_seats");
                    logger.trace("Available seats for trip {}: {}", tripId, availableSeats);
                    return availableSeats;
                }
            }
        }catch (Exception e) {
            logger.error("Error counting available seats for trip {}", tripId, e);
        }
        return -1;
    }

    @Override
    public Long add(Seat entity) {
        logger.info("Adding seat: {}", entity);

        String sql = "INSERT INTO seats (trip_id, reservation_id, number) VALUES (?, ?, ?)";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {

            statement.setLong(1, entity.getTripId());
            statement.setObject(2, entity.getReservationId(), Types.BIGINT);
            statement.setInt(3, entity.getNumber());

            int result = statement.executeUpdate();
            logger.trace("Seat inserted successfully: {}", result);

            try (ResultSet generatedKeys = statement.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    Long id = generatedKeys.getLong(1);
                    entity.setId(id);
                    return id;
                }
            }
            return null;
        } catch (Exception e) {
            logger.error("Error inserting seat {}", entity, e);
            return null;
        }
    }

    @Override
    public void delete(Long id) {
        logger.info("Deleting seat {}", id);

        String sql = "DELETE FROM seats WHERE id = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setLong(1, id);
            int result = statement.executeUpdate();

            logger.trace("Seat deleted successfully: {}", result);

        } catch (Exception e) {
            logger.error("Error deleting seat {}", id, e);
        }
    }

    @Override
    public void update(Seat entity) {
        logger.info("Updating seat {}", entity);

        String sql = "UPDATE seats SET trip_id=?, reservation_id=?, number=?, client_name =?  WHERE id=?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setLong(1, entity.getTripId());
            statement.setObject(2, entity.getReservationId(), Types.BIGINT);
            statement.setInt(3, entity.getNumber());
            statement.setObject(4, entity.getClientName(), Types.VARCHAR);
            statement.setLong(5, entity.getId());

            int result = statement.executeUpdate();
            logger.trace("Seat updated successfully: {}", result);

        } catch (Exception e) {
            logger.error("Error updating seat {}", entity, e);
        }
    }

    @Override
    public Seat findById(Long id) {
        logger.info("Finding seat {}", id);

        String sql = "SELECT * FROM seats WHERE id=?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setLong(1, id);

            try (ResultSet result = statement.executeQuery()) {
                if (result.next()) {
                    return buildSeatFromResultSet(result);
                }
            }

        } catch (Exception e) {
            logger.error("Error finding seat {}", id, e);
        }

        return null;
    }

    @Override
    public List<Seat> findAll() {
        logger.info("Finding all seats");

        String sql = "SELECT * FROM seats";
        List<Seat> seats = new ArrayList<>();

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql);
             ResultSet result = statement.executeQuery()) {

            while (result.next()) {
                seats.add(buildSeatFromResultSet(result));
            }

        } catch (Exception e) {
            logger.error("Error finding all seats", e);
        }

        return seats;
    }

    private Seat buildSeatFromResultSet(ResultSet result) throws SQLException {

        Long id = result.getLong("id");
        int number = result.getInt("number");
        Long reservationId = result.getLong("reservation_id");
        if (result.wasNull()) {
            reservationId = null;
        }
        Long tripId = result.getLong("trip_id");
        String clientName = result.getString("client_name");
        if(clientName == null)
            clientName = "-";


        return new Seat(id, number, reservationId, tripId, clientName);
    }
}

